dns.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"github.com/miekg/dns"
+	"github.com/prometheus/common/log"
+)
+
+// validRRs checks a slice of RRs received from the server against a DNSRRValidator.
+func validRRs(rrs *[]dns.RR, v *DNSRRValidator) bool {
+	// Probe should fail if a DNSRRValidator contains some rules, however there are no
+	// RRs of a given type in the response.
+	if len(*rrs) == 0 {
+		return len(v.FailIfMatchesRegexp) == 0 && len(v.FailIfNotMatchesRegexp) == 0
```

Comment:
Returning false would be the designed behaviour in this case, as described by the comment. The assumption here is that if you specify `FailIfMatchesRegexp` or `FailIfNotMatchesRegexp`, you expect at least a single RR to be returned (to actually do the matching).

Thinking more about this, I agree that this might not be the most obvious behaviour.

I've changed this to fail only when `FailIfNotMatchesRegexp` is specified, but no RRs are returned, which I believe might be more intuitive.